prov/shm/src/smr_progress.c

```diff
@@ -643,83 +627,92 @@ static int smr_progress_inject_atomic(struct smr_cmd *cmd, struct fi_ioc *ioc,
 
 	switch (cmd->msg.hdr.op_src) {
 	case smr_src_inline:
-		entry->err = smr_progress_inline(cmd, entry->iface, entry->device,
-						 entry->iov, entry->iov_count,
-						 &total_len);
+		err = smr_progress_inline(cmd, iface, device,
+					  rx_entry->iov, rx_entry->count,
+					  &total_len);
 		ep->region->cmd_cnt++;
 		break;
 	case smr_src_inject:
-		entry->err = smr_progress_inject(cmd, entry->iface, entry->device,
-						 entry->iov, entry->iov_count,
-						 &total_len, ep, 0);
+		err = smr_progress_inject(cmd, iface, device,
+					  rx_entry->iov, rx_entry->count,
+					  &total_len, ep, 0);
 		ep->region->cmd_cnt++;
 		break;
 	case smr_src_iov:
-		entry->err = smr_progress_iov(cmd, entry->iov, entry->iov_count,
-					      &total_len, ep, 0);
+		err = smr_progress_iov(cmd, rx_entry->iov, rx_entry->count,
+				       &total_len, ep, 0);
 		break;
 	case smr_src_mmap:
-		entry->err = smr_progress_mmap(cmd, entry->iface, entry->device,
-					       entry->iov, entry->iov_count,
-					       &total_len, ep);
+		err = smr_progress_mmap(cmd, iface, device,
+					rx_entry->iov, rx_entry->count,
+					&total_len, ep);
 		break;
 	case smr_src_sar:
-		sar = smr_progress_sar(cmd, entry, entry->iface, entry->device,
-				       entry->iov, entry->iov_count, &total_len, ep);
+		sar = smr_progress_sar(cmd, rx_entry, iface, device,
+				       rx_entry->iov, rx_entry->count, &total_len, ep);
 		break;
 	case smr_src_ipc:
-		entry->err = smr_progress_ipc(cmd, entry->iface, entry->device,
-					      entry->iov, entry->iov_count,
-					      &total_len, ep, 0);
+		err = smr_progress_ipc(cmd, iface, device,
+				       rx_entry->iov, rx_entry->count,
+				       &total_len, ep, 0);
 		break;
 	default:
 		FI_WARN(&smr_prov, FI_LOG_EP_CTRL,
 			"unidentified operation type\n");
-		entry->err = -FI_EINVAL;
-	}
-
-	comp_buf = entry->iov[0].iov_base;
-	comp_flags = (cmd->msg.hdr.op_flags | entry->flags) & ~SMR_MULTI_RECV;
-
-	if (entry->flags & SMR_MULTI_RECV) {
-		free_entry = smr_progress_multi_recv(ep, entry, total_len);
-		if (free_entry) {
-			comp_flags |= SMR_MULTI_RECV;
-			if (sar)
-				sar->rx_entry.flags |= SMR_MULTI_RECV;
-		}
+		err = -FI_EINVAL;
 	}
 
+	comp_buf = rx_entry->iov[0].iov_base;
+	comp_flags = smr_rx_cq_flags(cmd->msg.hdr.op, rx_entry->flags,
+				     cmd->msg.hdr.op_flags);
 	if (!sar) {
-		ret = smr_complete_rx(ep, entry->context, cmd->msg.hdr.op,
+		if (err) {
+			FI_WARN(&smr_prov, FI_LOG_EP_CTRL,
+				"error processing op\n");
+			ret = smr_write_err_comp(ep->util_ep.rx_cq, rx_entry->context,
+						 comp_flags, rx_entry->tag,
+						 err);
+		} else {
+			ret = smr_complete_rx(ep, rx_entry->context, cmd->msg.hdr.op,
 				comp_flags, total_len, comp_buf, cmd->msg.hdr.id,
-				cmd->msg.hdr.tag, cmd->msg.hdr.data, entry->err);
+				cmd->msg.hdr.tag, cmd->msg.hdr.data);
+		}
 		if (ret) {
 			FI_WARN(&smr_prov, FI_LOG_EP_CTRL,
 				"unable to process rx completion\n");
 		}
+		ep->srx->owner_ops->free_entry(rx_entry);
```

Comment:
should this be outside the if (!sar) statement?
if sar is not NULL, I don't see where the rx_entry will be freed.

Reply:
Kept this outside the if (!sar) because we can't free the entry until the peer is done and has completed that operation. But you were right, it was missing the free on the delayed sar case in progress_sar_list. Added it over there.